### ISSUE TITLE
add a way to tell if the content type has changed on the response

### ISSUE THREAD

**Issue by mmerickel** [Pylons/webob#204]
In Pyramid our renderers attempt to change the `content_type` of the `response` object to a sane default **if and only if** the user has not changed it themselves. We've traditionally used `if response.content_type == response.default_content_type` where the `default_content_type == 'text/html'` however in some cases we've found the user attempting to override a renderer like the `string` renderer to return `text/html`. In this case the user sets `response.content_type = 'text/html'` and the `string` renderer sees no changes and overrides it to `text/plain`.

TLDR A way to tell if the `content_type` was set.

Related: https://github.com/Pylons/pyramid/pull/1565 and https://github.com/Pylons/pyramid/issues/1344

**Comment by bertjwregeer** [comment]
I really want to remove the default of setting the `content_type` upon `Response` creation.

Here's what I am thinking:

- Remove the default content type addition, it is now up to the user to correctly set the content type
- Add a warning if `Response` is turned into an actual WSGI response (in `__call__`) and there is no content type set
- Make sure that content type is NOT set on certain response types that don't allow a body

This would be a backwards compatible change, to start off with I was thinking of making the following changes:

- Remove the default content type addition, it is now up to the user to correctly set the content type
- Add the default content type if `Response` is turned into an actual WSGI response and no content type is set in the headers when in `__call__`
- Warn if a content type is set on certain responses that should not have a body and thus no content type.

**Comment by mmerickel** [comment]
It doesn't sound bw-compat to me to focus only on the case where the response is turned into a wsgi response. There are likely a lot of people using the response object in tests, etc without performing this conversion.

**Comment by bertjwregeer** [comment]
That's my worry too. I don't think it is necessarily a good idea for there to be a default content type, but I am not sure how to reconcile the two...

**Comment by mmerickel** [comment]
I think figuring out how to pass a `str` subclass marker through from `__init__` should be enough to track it, maybe. But I think it's tricky to get that value shuffled through untouched.

**Comment by bertjwregeer** [comment]
There are many places or ways to set the content type without going through one particular path in the code. That's what makes this so much more difficult.
